src/clr-angular/utils/focus/key-focus/key-focus.spec.ts

```diff
@@ -0,0 +1,242 @@
+    });
+
+    it('move focus with home/end keys', () => {
+      openMenu();
+      expect(document.activeElement.textContent).toBe('Button 1');
+      keyPress('End');
+      expect(document.activeElement.textContent).toBe('Button 3');
+      keyPress('Home');
+      expect(document.activeElement.textContent).toBe('Button 1');
+    });
+
+    it('no focus rotation', () => {
+      openMenu();
+      expect(document.activeElement.textContent).toBe('Button 1');
+      keyPress('ArrowUp');
+      expect(document.activeElement.textContent).toBe('Button 1');
+      keyPress('End');
+      expect(document.activeElement.textContent).toBe('Button 3');
+      keyPress('ArrowDown');
+      expect(document.activeElement.textContent).toBe('Button 3');
+    });
+
+    describe('direction vertical', () => {
+      it('move focus with arrow up/down keys', () => {
+        openMenu();
+        expect(document.activeElement.textContent).toBe('Button 1');
+        // Arrow down
+        keyPress('ArrowDown');
+        expect(document.activeElement.textContent).toBe('Button 2');
+        // Make sure it's only ArrowDown/Up
+        keyPress('ArrowRight');
+        expect(document.activeElement.textContent).toBe('Button 2');
+        keyPress('ArrowLeft');
+        expect(document.activeElement.textContent).toBe('Button 2');
+        // Arrow up
+        keyPress('ArrowUp');
+        expect(document.activeElement.textContent).toBe('Button 1');
+      });
+    });
+
+    describe('direction horizontal', () => {
+      beforeEach(() => {
+        component.direction = 'horizontal';
+      });
+
+      it('move focus with arrow left/right keys', () => {
+        openMenu();
+        expect(document.activeElement.textContent).toBe('Button 1');
+        // Arrow right
+        keyPress('ArrowRight');
+        expect(document.activeElement.textContent).toBe('Button 2');
+        // Make sure it's only ArrowRight/Left
+        keyPress('ArrowDown');
+        expect(document.activeElement.textContent).toBe('Button 2');
+        keyPress('ArrowUp');
+        expect(document.activeElement.textContent).toBe('Button 2');
+        // Arrow up
+        keyPress('ArrowLeft');
+        expect(document.activeElement.textContent).toBe('Button 1');
+      });
+    });
+
+    describe('direction both', () => {
+      beforeEach(() => {
+        component.direction = 'both';
+      });
+
+      it('move focus with all arrow keys', () => {
+        openMenu();
+        expect(document.activeElement.textContent).toBe('Button 1');
+        keyPress('ArrowRight');
+        expect(document.activeElement.textContent).toBe('Button 2');
+        keyPress('ArrowDown');
+        expect(document.activeElement.textContent).toBe('Button 3');
+        keyPress('ArrowUp');
+        expect(document.activeElement.textContent).toBe('Button 2');
+        keyPress('ArrowLeft');
+        expect(document.activeElement.textContent).toBe('Button 1');
+      });
+    });
+  });
+
+  describe('API', () => {
```

Comment:
In general I think there should be three suites in a component like this:

1) `TypeScript API` which covers things like subscriptions to services and the side effects of service related events (prolly n/a to this directive)
2) `Template API` that covers the @Inputs, @Outputs and @HostListeners along with any public properties and methocs that make up the consumer API
3) `View` or `View Basics` that covers host and DOM element classes, DOM structure, class changes based on component state etc.

It would also help to see the tests for @HostListeners grouped/organized by event type.

Reply:
done